### DISCUSSION TITLE
2023/01/01/solving_n_plus_1_queries_on_rails

### DISCUSSION THREAD

**Discussion by giscus** [ananthakumaran/ananthakumaran.github.com#4] [Announcements]
# 2023/01/01/solving_n_plus_1_queries_on_rails

https://ananthakumaran.in/2023/01/01/solving_n_plus_1_queries_on_rails.html

**Comment by simonw** [comment]
These days it's possible to write a SQL query that returns data from related - including many-to-many tables - in a single go using JSON aggregation functions.

Here's an example relating to this post: https://gist.github.com/simonw/3d6cbcd55beda108a88265a80f042726?permalink_comment_id=4421809#gistcomment-4421809

I wrote more about this technique here: https://til.simonwillison.net/sqlite/related-rows-single-query

**Reply by ananthakumaran** [reply]
This is a nice solution for this problem and definitely makes sense for simpler use cases. But the question of composability still remains. It's not possible to just write a single query per controller/action, as usually multiple business logic rules are involved, and translating all those to SQL is non trivial.

**Reply by brandonvmoore** [reply]
If it's an option, simonw's solution is absolutely the way to go.
